### ISSUE TITLE
swarm-listener keeps sending create command even when the service was removed

### ISSUE THREAD

**Issue by galindro** [vfarcic/docker-flow-swarm-listener#29]
@vfarcic 

First of all, thank you for your great job writing the Devops 2.1 book. It openned my mind and now I'm really prepared to build a true swarm cluster.

I've found a bug in swarm-listener service. It is keeping sending create commands to docker-flow-proxy service even after remove a service / stack from docker. It is ocurring when there is a failed instruction passed a given service creation (in my case docker registry). So, the tasks keeps restarting and restarting. If I remove the service, the swarm-listener doesn't identify that the service was removed and keeps executing the create command. IMHO, I think that those commands are in a kind of queue in memory and they are executed even after the service is down.

I've make a [video](https://youtu.be/ojvnALXuzn0) to show you the bug. The files that I've used are these:

```bash
# docker-compose-global.yml
version: "3.2"

secrets:
  cert-i.sb.pem:
    file: ./cert-i.sb.pem
```
```bash
version: "3.2"

networks:
  out:
    driver: overlay
    ipam:
      driver: default
      config:
        - subnet: 10.100.0.0/16

secrets:
  global_cert-i.sb.pem:
    external: true

services:
  proxy:
    image: vfarcic/docker-flow-proxy
    deploy:
      mode: global
      placement:
        constraints:
          - engine.labels.proxy == true
          - engine.labels.availability_zone == sa-east-1c
    environment:
      LISTENER_ADDRESS: "swarm-listener"
      MODE: "swarm"
    secrets:
      - source: global_cert-i.sb.pem
        target: cert-i.sb.pem
    networks:
      - out
    ports:
      - "443:443"
      - "80:80"
      - "8080:8080"

  swarm-listener:
    image: vfarcic/docker-flow-swarm-listener
    deploy:
      mode: global
      placement:
        constraints:
          - node.role == manager
    environment:
      DF_NOTIF_CREATE_SERVICE_URL: "http://proxy:8080/v1/docker-flow-proxy/reconfigure"
      DF_NOTIF_REMOVE_SERVICE_URL: "http://proxy:8080/v1/docker-flow-proxy/remove"
    networks:
      - out
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
```

```bash
version: "3.2"

networks:
  proxy_out:
    external: true
  in:
    driver: overlay
    ipam:
      driver: default
      config:
        - subnet: 10.101.0.0/16
    internal: true

secrets:
  global_cert-i.sb.pem:
    external: true

services:
  registry:
    image: registry:2
    deploy:
      placement:
        constraints:
          - engine.labels.infra == true
          - engine.labels.availability_zone == sa-east-1c
      replicas: 1
      labels:
        com.df.notify: "true"
        com.df.distribute: "true"
        com.df.servicePath: "/registry"
        com.df.reqPathReplace: "/"
        com.df.reqPathSearch: "/registry/"
        com.df.port: 5000
      resources:
        reservations:
          memory: 100m
    environment:
      REGISTRY_VERSION: 0.1
      
      # Here I'm setting a wrong path to the certificate file. 
      # This will cause an error in container and the service will keep restarting 
      REGISTRY_HTTP_TLS_CERTIFICATE: "/wrong_path/cert-i.sb.crt"
      REGISTRY_HTTP_TLS_KEY: "/wrong_path/cert-i.sb.key"
      
      REGISTRY_STORAGE: "s3"
      REGISTRY_STORAGE_S3_REGION: "sa-east-1"
      REGISTRY_STORAGE_S3_BUCKET: "docker.infra.registry.images"
    secrets:
      - source: global_cert-i.sb.pem
        target: cert-i.sb.pem
    networks:
      - proxy_out
      - in
```

**Comment by vfarcic** [comment]
You're right. Once a service is created it enters a loop trying to reconfigure the proxy. If that service is removed it will not be aware of it (in that loop). It will stop after a while (there is a timeout) but it's very confusing if you look at logs. It looks like is infinite loop even though it isn't.

I'll work on an improvement so that the loop is stopped if a service is removed. I hope to have it done by the end of the week.

**Comment by vfarcic** [comment]
The fix is available in the release 1.33.

When a service is removed, the loop that sends create service notifications will stop.

Can you please try it out and let me know if you find any issues.

**Comment by vfarcic** [comment]
It's available in the release 1.34. Can you please try it out and let me know if you experience any problems?
